3scale/APIcast
# Introduce APICAST_LUA_SOCKET_KEEPALIVE_REQUESTS environment variable

## What 

A patch for https://issues.redhat.com/browse/THREESCALE-11321

## Verification steps
* Check out this branch
* Create a apicast-config.json file with the following content
```
cat <<EOF >apicast-config.json
{
    "services": [
        {
            "id": "1",
            "backend_version": "1",
            "proxy": {
                "hosts": [
                    "one"
                ],
                "api_backend": "https://echo-api.3scale.net:443",
                "backend": {
                    "endpoint": "http://127.0.0.1:8081",
                    "host": "backend"
                },
                "policy_chain": [
                    {
                        "name": "apicast.policy.apicast"
                    }
                ],
                "proxy_rules": [
                    {
                        "http_method": "GET",
                        "pattern": "/",
                        "metric_system_name": "hits",
                        "delta": 1,
                        "parameters": [],
                        "querystring_parameters": {}
                    }
                ]
            }
        }
    ]
} 
EOF
```
* Start development environment
```
make development
make dependencies
```
* Run apicast locally with `APICAST_LUA_SOCKET_KEEPALIVE_REQUESTS=5`
```
THREESCALE_DEPLOYMENT_ENV=staging APICAST_LOG_LEVEL=debug APICAST_WORKER=1 APICAST_CONFIGURATION_LOADER=lazy APICAST_CONFIGURATION_CACHE=0 APICAST_LUA_SOCKET_KEEPALIVE_REQUESTS=5 THREESCALE_CONFIG_FILE=apicast-config.json ./bin/apicast
```
* Capture apicast IP
```
APICAST_IP=$(docker inspect apicast_build_0-development-1 | yq e -P '.[0].NetworkSettings.Networks.apicast_build_0_default.IPAddress' -)
```
* Send a request
```
curl -i -k -H "Host: one" -H "Accept: application/json" "http://${APICAST_IP}:8080/?user_key="
```
* Note the following line in the log
```
 proxy.lua:26: new(): connection to 127.0.0.1:8081 established, reused times: 1
```
* Keep sending more requests, after the 5 request, the reused time should reset to 0
```
proxy.lua:26: new(): connection to 127.0.0.1:8081 established, reused times: 0,

```
